src/EventListener/CreateSchemaListener.php

```diff
@@ -105,15 +112,29 @@ public function postGenerateSchemaTable(GenerateSchemaTableEventArgs $eventArgs)
         $revisionTable->addColumn($this->config->getRevisionFieldName(), $this->config->getRevisionIdFieldType());
         $revisionTable->addColumn($this->config->getRevisionTypeFieldName(), Types::STRING, ['length' => 4]);
         if (!\in_array($cm->inheritanceType, [ClassMetadata::INHERITANCE_TYPE_NONE, ClassMetadata::INHERITANCE_TYPE_JOINED, ClassMetadata::INHERITANCE_TYPE_SINGLE_TABLE], true)) {
-            throw new \Exception(\sprintf('Inheritance type "%s" is not yet supported', $cm->inheritanceType));
+            throw new \RuntimeException(\sprintf('Inheritance type "%s" is not yet supported', $cm->inheritanceType));
         }
 
-        $primaryKey = $entityTable->getPrimaryKey();
-        \assert(null !== $primaryKey);
-        $pkColumns = $primaryKey->getColumns();
-        $pkColumns[] = $this->config->getRevisionFieldName();
-        $revisionTable->setPrimaryKey($pkColumns);
-        $revIndexName = $this->config->getRevisionFieldName().'_'.md5($revisionTable->getName()).'_idx';
+        if ($this->isDbal4_3()) {
+            $primaryKey = $entityTable->getPrimaryKeyConstraint();
+            \assert(null !== $primaryKey);
+            $pkColumns = $primaryKey->getColumnNames();
+            /** @psalm-suppress ArgumentTypeCoercion */
+            $pkColumns[] = new UnqualifiedName(Identifier::unquoted($this->config->getRevisionFieldName())); // @phpstan-ignore-line
+            /** @psalm-suppress InternalMethod */
+            $revisionTable->addPrimaryKeyConstraint(new PrimaryKeyConstraint(null, $pkColumns, $primaryKey->isClustered())); // @phpstan-ignore-line
+        } else {
+            /** @psalm-suppress DeprecatedMethod */
+            $primaryKey = $entityTable->getPrimaryKey();
+            \assert(null !== $primaryKey);
+            /** @psalm-suppress DeprecatedMethod */
+            $pkColumns = $primaryKey->getColumns();
+            $pkColumns[] = $this->config->getRevisionFieldName();
+            /** @psalm-suppress DeprecatedMethod */
+            $revisionTable->setPrimaryKey($pkColumns);
+        }
+        /** @psalm-suppress InternalMethod */
```

Comment:
Is there a way to avoid the use of internal methods ?

If not, should we open an issue on doctrine side to remove the `@internal` ?